doc/design/refactor/multi_device.md

```diff
@@ -0,0 +1,47 @@
+# Design Doc: How to Execute the ProgramDesc on Multiple Device
+
+A sequence of optimizers will convert a ProgramDesc to ExecutionPlan, which allow
+the PaddlePaddle program running on multi devices such CPU, GPU or FPGA,
+or multi nodes.
+
+The interface of ExecutionPlan is as followed:
+
+```cpp
+enum ExecutionType {
+    SINGLE_THREAD = 0,
+    MULTIPLE_CPU = 1,
+    MULTIPLE_GPU = 2,
+    MULTIPLE_NODE = 3
+};
+
+class ExecutionPlan {
+  ExecutionType type_;
+  vector<ProgramDescBind*> trainers_;
+  vector<ProgramDescBind*> pservers_;
+};
+```
+
+## Running with Multi-CPU
+
+<img src="src/multi_cpu.png">
+
+1. Use Multi-CPU Optimizer to convert the ProgramDesc to ExecutionPlan
+
+    For the data parallelism, we need to pass the attribution `start` and `end`
+    index for the mini-batch, and this will be calculated in the optimizer step.
+1. `multiCPUExecutor` execute the ExecutionPlan which the type is Multi CPU
+
+    the `MultiCPUExecutor` will call `block.clone()` and `scope.clone()` to make
+    a list of blocks and scopes, the size equals the thread number, and then execute
+    the graph in each thread.
+1. Collect the gradients and update the parameters
```

Comment:
In multi-threads, we need only one copy parameter in CPU memory. It;s different from multi-GPUs, which each GPU will hold parameter in its own GPU memory.

Reply:
Got it, there would be one parameters and it could be on the global scope, for anther, the gradients for each thread is different, so maybe we also need to create the gradients on multi scope.